resolver.go

```diff
@@ -484,7 +484,7 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			resp, err = co.ReadMsg()
 			// Truncated DNS replies should be sent to the client so that the
 			// client can retry over TCP
-			if err != nil && !resp.Truncated {
+			if err != nil && (resp != nil && !resp.Truncated) {
```

Comment:
Should we move https://github.com/moby/libnetwork/pull/2540/files#diff-6756fdfda338c13b910a6416f2ac9f54R494-R497 before this instead?

Reply:
(this code):

```go
			if resp == nil {
				logrus.Debugf("[resolver] external DNS %s:%s returned empty response for %q", proto, extDNS.IPStr, name)
				break
			}
```

Reply:
oh, I guess not, because then we wouldn't have the `err`

Reply:
`forwardQueryEnd` still needs to be called if `err == nil && resp == nil`, I think. I was trying to shuffle the logic around as little as possible, but I maybe it's worth rewriting some of this?

Reply:
Yeah,  it's probably ok for now, but we might want to clean it up a bit in a follow-up